lucet-runtime/lucet-runtime-internals/src/alloc/tests.rs

```diff
@@ -820,6 +860,82 @@ macro_rules! alloc_tests {
                 Ok(_) => panic!("unexpected success"),
             }
         }
+
+        /// This test exercises custom limits on the heap_memory_size.
+        /// In this scenario, the Region has a limit on the heap
+        /// memory size, but the instance has a larger limit.  An
+        /// instance's custom limit must not exceed the Region's.
+        #[test]
+        fn reject_heap_memory_size_exeeds_region_limits() {
+            let region = TestRegion::create(1, &LIMITS).expect("region created");
+            let module = MockModuleBuilder::new()
+                .with_heap_spec(THREE_PAGE_MAX_HEAP)
+                .build();
+            let res = region
+                .new_instance_builder(module.clone())
+                .with_heap_size_limit(&LIMITS.heap_memory_size * 2)
+                .build();
+
+            match res {
+                Err(Error::InvalidArgument(
+                    "heap memory size requested for instance is larger than slot allows",
+                )) => (),
+                Err(e) => panic!("unexpected error: {}", e),
+                Ok(_) => panic!("unexpected success"),
+            }
+        }
+
+        /// This test exercises custom limits on the heap_memory_size.
+        /// In this scenario, successfully create a custom-sized
+        /// instance followed by a default-sized instance to affirm
+        /// that a custom size doesn't somehow overwrite the default size.
+        #[test]
+        fn custom_size_does_not_break_default() {
+            let region = TestRegion::create(2, &LIMITS).expect("region created");
+            let module = MockModuleBuilder::new()
+                .with_heap_spec(THREE_PAGE_MAX_HEAP)
+                .build();
+
+            // Build an instance that is has custom limits that are big
+            // enough to accommodate the HeapSpec.
+            let mut custom_inst = region
+                .new_instance_builder(module.clone())
+                .with_heap_size_limit(THREE_PAGE_MAX_HEAP.initial_size as usize)
+                .build()
+                .expect("new instance succeeds");
+
+            // Affirm that its heap is the expected size.
+            let heap_len = custom_inst.alloc().heap_len();
+            assert_eq!(heap_len, THREE_PAGE_MAX_HEAP.initial_size as usize);
+
+            // Build a default-sized instance, to make sure the custom limits
+            // didn't break the defaults.
+            let default_inst = region
+                .new_instance(module.clone())
+                .expect("new_instance succeeds");
+
+            // Affirm that its heap is the expected size.
+            let heap_len = default_inst.alloc().heap_len();
+            assert_eq!(heap_len, THREEPAGE_INITIAL_SIZE as usize);
```

Comment:
Ugh.  I don't like what I did here.  I am going to revise this to use a different initial size in the HeapSpec.  Please hold.